server/src/main/java/com/flipkart/varadhi/qos/RateLimiter.java

```diff
@@ -0,0 +1,228 @@
+package com.flipkart.varadhi.qos;
+
+
+import com.flipkart.varadhi.entities.ratelimit.LoadInfo;
+import com.flipkart.varadhi.entities.ratelimit.RateLimitReason;
+import com.flipkart.varadhi.entities.ratelimit.SuppressionData;
+import com.flipkart.varadhi.entities.ratelimit.TrafficData;
+import com.flipkart.varadhi.services.VaradhiTopicService;
+
+import com.flipkart.varadhi.utils.weights.WeightFunction;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Multiset;
+import com.google.common.collect.PeekingIterator;
+
+import com.google.common.collect.TreeMultiset;
+
+import java.util.*;
+import java.util.concurrent.*;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * - Initialise a priorityQueue (min heap) to maintain (time, order, data).
+ * - Custom comparator to sort by time asc, order desc.
+ *
+ * Thread 1:
+ * - pods will send data (from, to, data), break it into 2 events.
+ * - (from, 1, +data/(to-from)) and (to, 0, -data/(to-from)) and add into PQ if
+ * (from > (now() - Window size)) // skip adding if old data
+ *
+ * Thread 2: (pooling)
+ * - maintain aggreagte value to use in check breach.
+ * - peek and pop till now()
+ * - compute running avg.
+ * - check for RL. If true, add the suppression factor
+ */
+
+@Slf4j
+public class RateLimiter { // TODO(rl): impl publisher, for observer
```

Comment:
this class, needs better method documentation, and some  more class documentation around who is sending data to it and how. and how is suppression data being sent back.